usaspending_api/bulk_download/v2/views.py

```diff
@@ -0,0 +1,311 @@
+                                          Q(funding_agency_id=filters['sub_agency'])))
+
+        # Put it all together
+        table = value_mappings[award_level]['table']
+        if and_queryset_filters and or_queryset:
+            filtered_queryset = table.objects.filter(**and_queryset_filters).filter(or_queryset)
+        elif and_queryset_filters:
+            filtered_queryset = table.objects.filter(**and_queryset_filters)
+        elif or_queryset:
+            filtered_queryset = table.objects.filter(or_queryset)
+        else:
+            filtered_queryset = table.objects.all()
+
+        # Note: Ordering by date for testing/verification for now
+        #       If there's a request to sort by a specific attribute,
+        #       here's where it will be.
+        order_by = date_attribute if date_attribute else value_mappings[award_level]['action_date']
+        filtered_queryset = filtered_queryset.order_by(order_by)
+
+        return filtered_queryset
+
+
+    def get_csv_sources(self, json_request):
+        """
+        Generate the CSV Sources (filtered queryset and metadata) based on the request
+        """
+        award_levels = json_request['award_levels']
+        # TODO: Send use file_format
+        file_format = json_request['file_format']
+
+        csv_sources = []
+        if not isinstance(award_levels, list):
+            raise InvalidParameterException('award_levels parameter not provided as a list')
+        for award_level in award_levels:
+            if 'filters' in json_request:
+                queryset = self.process_filters(json_request['filters'], award_level)
+            else:
+                table = value_mappings[award_level]['table']
+                queryset = table.objects.all()
+            if award_level == 'prime_awards':
+                d1_source = csv_selection.CsvSource('award', 'd1')
+                d2_source = csv_selection.CsvSource('award', 'd2')
+                verify_requested_columns_available((d1_source, d2_source), json_request['columns'])
+                d1_source.queryset = queryset & Award.objects.filter(latest_transaction__contract_data__isnull=False)
+                d2_source.queryset = queryset & Award.objects.filter(latest_transaction__assistance_data__isnull=False)
+                csv_sources.extend([d1_source, d2_source])
+            elif award_level == 'sub_awards':
+                # NOT IMPLEMENTED
+                # d1_source = csv_selection.CsvSource('subaward', 'd1')
+                # d2_source = csv_selection.CsvSource('subaward', 'd2')
+                # verify_requested_columns_available((d1_source, d2_source), json_request['columns'])
+                raise NotImplementedError
+            else:
+                raise InvalidParameterException('Invalid parameter for award_levels: {}'.format(award_level))
+
+        return tuple(csv_sources)
+
+    DOWNLOAD_NAME = 'awards'
```

Comment:
Might be a good idea to move this to the top of the class, it could get lost down here

Reply:
Good call and moved to the top.